# EXPLORATORY DATA ANALYSIS on Apple Inc. Stock

What is APPLE?

Apple Inc. is an American multinational technology company that specializes in consumer electronics, computer software, and online services. Apple is the world's largest technology company by revenue and, since January 2021, the world's most valuable company.


Information about this dataset

This dataset provides historical data of APPLE INC. stock (AAPL). The data is available at a daily level. Currency is USD.

Here is one row for each day as indicated in the date field:

*open and close : Opening price when market opens, and closing prices when market closes, of stock on the particular day.

*high and low : High stock price, and low stock price on that day.

*adj close : The adjusted closing price amends a stock's closing price to reflect that stock's value after accounting for any corporate actions such as dividend payouts, stock splits, or the issuance of more shares. It is often used when examining historical returns or doing a detailed analysis of past performance.

*volume : Number of shares traded on that day

Credit source for additional reading: 

https://corporatefinanceinstitute.com/resources/knowledge/trading-investing/adjusted-closing-price/

https://pubs.opengroup.org/onlinepubs/007908799/xsh/strftime.html

https://davidkeya.com/portfolio/Sql_Pandas_EDA/

## SQLite Jupyter Notebook: Application of Magic Functions

Using jupyter notebook, we use magic functions under sqlite to query sql.

Below query list all magic functions in available in jupyter notebook.

In [36]:
%lsmagic

Available line magics:
%alias  %alias_magic  %autoawait  %autocall  %automagic  %autosave  %bookmark  %cd  %clear  %cls  %colors  %conda  %config  %connect_info  %copy  %ddir  %debug  %dhist  %dirs  %doctest_mode  %echo  %ed  %edit  %env  %gui  %hist  %history  %killbgscripts  %ldir  %less  %load  %load_ext  %loadpy  %logoff  %logon  %logstart  %logstate  %logstop  %ls  %lsmagic  %macro  %magic  %matplotlib  %mkdir  %more  %notebook  %page  %pastebin  %pdb  %pdef  %pdoc  %pfile  %pinfo  %pinfo2  %pip  %popd  %pprint  %precision  %prun  %psearch  %psource  %pushd  %pwd  %pycat  %pylab  %qtconsole  %quickref  %recall  %rehashx  %reload_ext  %ren  %rep  %rerun  %reset  %reset_selective  %rmdir  %run  %save  %sc  %set_env  %sql  %store  %sx  %system  %tb  %time  %timeit  %unalias  %unload_ext  %who  %who_ls  %whos  %xdel  %xmode

Available cell magics:
%%!  %%HTML  %%SVG  %%bash  %%capture  %%cmd  %%debug  %%file  %%html  %%javascript  %%js  %%latex  %%markdown  %%perl  %%prun  %%pypy  %%p

Credit source for additional reading: https://www.tutorialspoint.com/jupyter/ipython_magic_commands.htm

Connecting to Database:

At this point we’re almost ready to run sql scripts from Python code cells in either Spyder or Jupyter but we need to establish a connection to the database first. This can be done in several ways depending on your database choice and file or server location. Here are some examples:

-sqlite:// –> This will establish a temp connection with sqlite.

-sqlite:////Users/ABC/Desktop/yourdatabase.sqlite –> This can be used to specify a database in a specific location. Also if the database file doesn’t exist it will be initiated.

-sqlite:///yourdatabase.sqlite –> This can be used to connect to a database file in the same working directory.


Alternative Connection Options:

IBM DB2 Cloud Database Connection Example:

‘%sql ibm_db_sa://etd48682:S5d4Wg6s3%6211n@dashdb-txn-sbox-yp-dal09-08.services.dal.bluemix.net:50000/BLUDB

MySQL Database Connection Example:

mysql://user:pwd@localhost/database_name

Oracle Database Connection Example:

oracle://user:pwd@127.0.0.1:1962/database_name

PostgreSQL Database Connection Examples:

postgres://name:pwd@localhost:6259/database_name

postgres://localhost:6259/database_name


at this point sql magic is connected to the database and ready to use. It’s an impressive solution to using sql from Ipython console in Jupyter Notebook. What most people might not know is that Ipython also runs in Spyder IDE. Spyder is a great IDE that’s used by scientists, researchers and data scientists as well.

Source: https://holypython.com/python-sqlite-tutorial/python-magic-for-sql/

### Changing file format to db/sqlite file to query database using SQLite query

Since the raw dataset in csv file, we change it to db file format to query the database with SQL query.

In [2]:
import sqlite3
import pandas as pd

In [3]:
conn = sqlite3.connect('applestock.db') #creating applestock database

In [4]:
app = pd.read_csv('AAPL.csv')

In [5]:
res = conn.execute("SELECT name FROM sqlite_master WHERE type='table';")
for name in res:
    print(name[0])

AppleStock


To add SQL extension, we use the query:

In [1]:
%load_ext sql

Connect to applestock database with query:

In [2]:
%sql sqlite:///applestock.db

Now, we would like to view the table with its field from this dataset, using SQLite query below:

In [3]:
%sql SELECT sql from sqlite_master WHERE type='table';

 * sqlite:///applestock.db
Done.


[('CREATE TABLE "AppleStock" (\n"index" INTEGER,\n  "Date" TEXT,\n  "Open" REAL,\n  "High" REAL,\n  "Low" REAL,\n  "Close" REAL,\n  "Adj Close" REAL,\n  "Volume" INTEGER\n)',)]

We can see that the table in this dataset named AppleStock.

There are 8 fields in this table:

*Index

*Date

*Open

*High

*Low

*Close

*Adj Close

*Volume

The explanation of every column/field can be found on the early discription of this notebook.

# Performing Exploratory Data Analysis (EDA)

Now, we would like to view the table in this dataset using query below.

In [4]:
%%sql
SELECT *
FROM AppleStock
LIMIT 10;

 * sqlite:///applestock.db
Done.


[(0, '1980-12-12', 0.128348, 0.128906, 0.128348, 0.128348, 0.100453, 469033600),
 (1, '1980-12-15', 0.12220999999999999, 0.12220999999999999, 0.121652, 0.121652, 0.095213, 175884800),
 (2, '1980-12-16', 0.11328099999999999, 0.11328099999999999, 0.112723, 0.112723, 0.088224, 105728000),
 (3, '1980-12-17', 0.115513, 0.11607100000000001, 0.115513, 0.115513, 0.090408, 86441600),
 (4, '1980-12-18', 0.118862, 0.11942, 0.118862, 0.118862, 0.093029, 73449600),
 (5, '1980-12-19', 0.126116, 0.126674, 0.126116, 0.126116, 0.098707, 48630400),
 (6, '1980-12-22', 0.13225399999999998, 0.13281300000000001, 0.13225399999999998, 0.13225399999999998, 0.10351099999999999, 37363200),
 (7, '1980-12-23', 0.13783499999999999, 0.13839300000000002, 0.13783499999999999, 0.13783499999999999, 0.107879, 46950400),
 (8, '1980-12-24', 0.14508900000000002, 0.145647, 0.14508900000000002, 0.14508900000000002, 0.11355599999999999, 48003200),
 (9, '1980-12-26', 0.15848199999999998, 0.15904000000000001, 0.15848199999999998, 0.15848199999999998, 0.12403800000000001, 55574400)]

%%sql

/*use this query to query the dataset from table applestock*/

SELECT *

FROM AppleStock;

Note that the comments we can insert in sqlite query, must be positioned after '%sql' or '%%sql' using symbol:

/* : as opening comment section in SQL query

*/ : as closing comment section in SQL query

Now, let us find number of occurences of shares trading from year 1980 till year 2021 using COUNT() function.

In [5]:
%%sql
/*Total shares trading occurences using COUNT() function*/
SELECT COUNT(*) AS Total_Occur_Shares_Traded
FROM AppleStock;

 * sqlite:///applestock.db
Done.


[(10336,)]

Thus, the number of events of shares traded by Apple Inc takes place is 10,336 times from year 1980 till year 2021.

Note that:
'AS' or 'as' stands for alias in SQL query to give field meaningful name based on our understanding of querying data.

Then, we can find total number of shares traded between year 1980 till 2021, using aggregate function SUM ().

In [6]:
%%sql
/*To get total number of shares Apple had traded from the year 1980 till year 2021*/
SELECT SUM (Volume) AS Total_Num_Shares_Traded
FROM AppleStock;

 * sqlite:///applestock.db
Done.


[(3450495313500,)]

We can see about total of 3.45 trillion number of shares traded by Apples Inc. from the year 1980 till latest year 2021.

Now, calculating total of market capitalization of Apple Inc shares trading history:

In [7]:
%%sql
/*To get total market capitalization of Apple from the year 1980 till year 2021*/
SELECT ROUND(SUM (Volume * [Adj Close]),2) AS Total_Market_Capitalization_AppleInc
FROM AppleStock;

 * sqlite:///applestock.db
Done.


[(24567336642411.16,)]

Apple Inc set records of history for total market capitalization with $ 24.567 trillion USD.

Querying market capitalization of Apple Inc shares trading history, based on respective year:

In [8]:
%%sql
SELECT strftime('%Y', Date) AS Year, ROUND(SUM (Volume * [Adj Close]),2) AS Market_Capitalization_AppleInc
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 138349567.55),
 ('1981', 688212548.55),
 ('1982', 1594793131.97),
 ('1983', 5367306965.82),
 ('1984', 3990725460.47),
 ('1985', 3371101517.06),
 ('1986', 6054551590.24),
 ('1987', 15349871255.97),
 ('1988', 12066337327.56),
 ('1989', 14821774414.84),
 ('1990', 12081670206.47),
 ('1991', 22179428670.12),
 ('1992', 16513769934.28),
 ('1993', 17085138770.35),
 ('1994', 14756296269.03),
 ('1995', 22804038704.73),
 ('1996', 10478569290.87),
 ('1997', 10731995758.56),
 ('1998', 27684322716.02),
 ('1999', 60512975132.12),
 ('2000', 76189072819.39),
 ('2001', 29539228316.43),
 ('2002', 22942369872.84),
 ('2003', 20056368712.34),
 ('2004', 74791761032.99),
 ('2005', 255014498236.06),
 ('2006', 465082004395.58),
 ('2007', 988402731718.04),
 ('2008', 1200086738679.55),
 ('2009', 613497602027.66),
 ('2010', 1179042854103.23),
 ('2011', 1387416314030.77),
 ('2012', 2339064692458.73),
 ('2013', 1502370539078.33),
 ('2014', 1282280395080.3),
 ('2015', 1424356439157.58),
 ('2016', 934353519283.76),
 ('2017', 974031534804.9),
 ('2018', 1541620217816.7),
 ('2019', 1421724219547.87),
 ('2020', 3653041355098.71),
 ('2021', 2904160956906.83)]

From the findings, Apple Inc has set the record of its market capitalization to be $ 2.904 trillion USD towards the end of year 2021.

Credit source additional reading:https://www.merrilledge.com/article/company-size-why-market-capitalization-matters-ose

https://www.investopedia.com/ask/answers/122314/what-difference-between-market-capitalization-and-market-value.asp#:~:text=Market%20capitalization%20is%20essentially%20a,and%20multiples%20the%20analyst%20uses.

We will try to count the number of occurences of shares trading for every year starting from 1980 until 2021, group by year.

In [9]:
%%sql
SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 13),
 ('1981', 253),
 ('1982', 253),
 ('1983', 253),
 ('1984', 253),
 ('1985', 252),
 ('1986', 253),
 ('1987', 253),
 ('1988', 253),
 ('1989', 252),
 ('1990', 253),
 ('1991', 253),
 ('1992', 254),
 ('1993', 253),
 ('1994', 252),
 ('1995', 252),
 ('1996', 254),
 ('1997', 253),
 ('1998', 252),
 ('1999', 252),
 ('2000', 252),
 ('2001', 248),
 ('2002', 252),
 ('2003', 252),
 ('2004', 252),
 ('2005', 252),
 ('2006', 251),
 ('2007', 251),
 ('2008', 253),
 ('2009', 252),
 ('2010', 252),
 ('2011', 252),
 ('2012', 250),
 ('2013', 252),
 ('2014', 252),
 ('2015', 252),
 ('2016', 252),
 ('2017', 251),
 ('2018', 251),
 ('2019', 252),
 ('2020', 253),
 ('2021', 236)]

Now, using SUM() funtion to get volume, which is number of shares traded on that year.

In [10]:
%%sql
/*Use SUM function to get total number of shares traded on respective year*/
SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade, SUM (Volume) AS Total_Shares_Traded
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 13, 1344851200),
 ('1981', 253, 8196944000),
 ('1982', 253, 21365008000),
 ('1983', 253, 44513011200),
 ('1984', 253, 41979033600),
 ('1985', 252, 45492272000),
 ('1986', 253, 53323222400),
 ('1987', 253, 59771308800),
 ('1988', 253, 41292977600),
 ('1989', 252, 50905825600),
 ('1990', 253, 44401940800),
 ('1991', 253, 57347651200),
 ('1992', 254, 41137913600),
 ('1993', 253, 56452928000),
 ('1994', 252, 57155896000),
 ('1995', 252, 74266539200),
 ('1996', 254, 53194220800),
 ('1997', 253, 71963360000),
 ('1998', 252, 115194195200),
 ('1999', 252, 137102705600),
 ('2000', 252, 120301596800),
 ('2001', 248, 94657796800),
 ('2002', 252, 77013927200),
 ('2003', 252, 71230252800),
 ('2004', 252, 121801668800),
 ('2005', 252, 182400982400),
 ('2006', 251, 215698966000),
 ('2007', 251, 246995985600),
 ('2008', 253, 285981206000),
 ('2009', 252, 143253686800),
 ('2010', 252, 151024927200),
 ('2011', 252, 124059339600),
 ('2012', 250, 131964204400),
 ('2013', 252, 102421569600),
 ('2014', 252, 63657952400),
 ('2015', 252, 52264199600),
 ('2016', 252, 38729911200),
 ('2017', 251, 27243106000),
 ('2018', 251, 34156144800),
 ('2019', 252, 28254942800),
 ('2020', 253, 39863855600),
 ('2021', 236, 21117286300)]

Now, try to compare closing price and adjusted closing price of shares within respective year.

Use ROUND () function to round off the query result to desired decimal place. In this query, only desired decimal place is 2.

For querying column / field name that has space between words, eg: Adj Close (there's gap between words 'Adj' and 'Close'), thus must use:

[ ] - for dealing with column / field name that has space between words.

In [11]:
%%sql
/*Comparing closing price and adjusted closing price every year*/
SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade, ROUND(SUM (Close),2) AS Close_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 13, 1.77, 1.38),
 ('1981', 253, 27.46, 21.49),
 ('1982', 253, 21.62, 16.92),
 ('1983', 253, 42.32, 33.12),
 ('1984', 253, 30.27, 23.69),
 ('1985', 252, 22.74, 17.8),
 ('1986', 253, 36.66, 28.69),
 ('1987', 253, 87.98, 68.99),
 ('1988', 253, 93.83, 74.1),
 ('1989', 252, 93.73, 74.73),
 ('1990', 253, 84.85, 68.38),
 ('1991', 253, 118.58, 96.6),
 ('1992', 254, 124.28, 102.13),
 ('1993', 253, 92.68, 76.85),
 ('1994', 252, 76.68, 64.58),
 ('1995', 252, 91.22, 77.74),
 ('1996', 254, 56.51, 48.45),
 ('1997', 253, 40.59, 34.8),
 ('1998', 252, 68.77, 58.96),
 ('1999', 252, 129.98, 111.44),
 ('2000', 252, 205.7, 176.36),
 ('2001', 248, 89.54, 76.77),
 ('2002', 252, 86.13, 73.84),
 ('2003', 252, 83.45, 71.55),
 ('2004', 252, 159.87, 137.07),
 ('2005', 252, 420.08, 360.17),
 ('2006', 251, 634.77, 544.23),
 ('2007', 251, 1149.88, 985.87),
 ('2008', 253, 1282.88, 1099.9),
 ('2009', 252, 1321.33, 1132.86),
 ('2010', 252, 2338.58, 2005.02),
 ('2011', 252, 3276.04, 2808.76),
 ('2012', 250, 5143.3, 4420.39),
 ('2013', 252, 4253.71, 3729.39),
 ('2014', 252, 5812.67, 5219.4),
 ('2015', 252, 7562.52, 6906.77),
 ('2016', 252, 6590.05, 6143.41),
 ('2017', 251, 9447.08, 8972.91),
 ('2018', 251, 11863.1, 11439.36),
 ('2019', 252, 13120.12, 12851.15),
 ('2020', 253, 24122.81, 23898.75),
 ('2021', 236, 32713.11, 32618.07)]

Let us compare the decrement of closing price and adjusted closing price within respective year using format percentage.

Since SQLite in Jupyter Notebook does not support CONCAT() function of standard SQL query, we use '||' to concatenate result with strings.

Credit source additional reading: https://www.sqlitetutorial.net/sqlite-string-functions/sqlite-concat/

In [12]:
%%sql
/*Decrement percentage from closing price to adjusted closing price within respective year*/
SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade,
       ROUND(SUM (Close),2) AS Close_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst,
       ROUND(((SUM(Close)-SUM([Adj Close])) / SUM(Close))*100,2) || ' ' || '%' AS Percent_CP_Decrement
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 13, 1.77, 1.38, '21.73 %'),
 ('1981', 253, 27.46, 21.49, '21.73 %'),
 ('1982', 253, 21.62, 16.92, '21.73 %'),
 ('1983', 253, 42.32, 33.12, '21.73 %'),
 ('1984', 253, 30.27, 23.69, '21.73 %'),
 ('1985', 252, 22.74, 17.8, '21.73 %'),
 ('1986', 253, 36.66, 28.69, '21.73 %'),
 ('1987', 253, 87.98, 68.99, '21.58 %'),
 ('1988', 253, 93.83, 74.1, '21.03 %'),
 ('1989', 252, 93.73, 74.73, '20.28 %'),
 ('1990', 253, 84.85, 68.38, '19.41 %'),
 ('1991', 253, 118.58, 96.6, '18.54 %'),
 ('1992', 254, 124.28, 102.13, '17.83 %'),
 ('1993', 253, 92.68, 76.85, '17.08 %'),
 ('1994', 252, 76.68, 64.58, '15.78 %'),
 ('1995', 252, 91.22, 77.74, '14.77 %'),
 ('1996', 254, 56.51, 48.45, '14.26 %'),
 ('1997', 253, 40.59, 34.8, '14.26 %'),
 ('1998', 252, 68.77, 58.96, '14.26 %'),
 ('1999', 252, 129.98, 111.44, '14.26 %'),
 ('2000', 252, 205.7, 176.36, '14.26 %'),
 ('2001', 248, 89.54, 76.77, '14.26 %'),
 ('2002', 252, 86.13, 73.84, '14.26 %'),
 ('2003', 252, 83.45, 71.55, '14.26 %'),
 ('2004', 252, 159.87, 137.07, '14.26 %'),
 ('2005', 252, 420.08, 360.17, '14.26 %'),
 ('2006', 251, 634.77, 544.23, '14.26 %'),
 ('2007', 251, 1149.88, 985.87, '14.26 %'),
 ('2008', 253, 1282.88, 1099.9, '14.26 %'),
 ('2009', 252, 1321.33, 1132.86, '14.26 %'),
 ('2010', 252, 2338.58, 2005.02, '14.26 %'),
 ('2011', 252, 3276.04, 2808.76, '14.26 %'),
 ('2012', 250, 5143.3, 4420.39, '14.06 %'),
 ('2013', 252, 4253.71, 3729.39, '12.33 %'),
 ('2014', 252, 5812.67, 5219.4, '10.21 %'),
 ('2015', 252, 7562.52, 6906.77, '8.67 %'),
 ('2016', 252, 6590.05, 6143.41, '6.78 %'),
 ('2017', 251, 9447.08, 8972.91, '5.02 %'),
 ('2018', 251, 11863.1, 11439.36, '3.57 %'),
 ('2019', 252, 13120.12, 12851.15, '2.05 %'),
 ('2020', 253, 24122.81, 23898.75, '0.93 %'),
 ('2021', 236, 32713.11, 32618.07, '0.29 %')]

From table above, we can see that the percent in decrement of closing price to closing price adjusted decreases as years go by.

We can see the trend that from year;

| Range Year | Percent Decrement of closing price |
|------------|------------------------------------|
|1980 - 1993 |        21.73% - 17.08%             |
|1994 - 2014 |        15.78% - 10.21%             |
|2015 - 2019 |         8.67% -  2.05%             |
|2020 - 2021 |         0.93% -  0.29%             |

Based on the source; https://corporatefinanceinstitute.com/resources/knowledge/trading-investing/adjusted-closing-price/

# Key takeaways:

* The adjusted closing price is a calculation adjustment made to a stock’s closing price.
* It is more complex and accurate than the closing price.
* The adjustment made to the closing price displays the true price of the stock because outside factors could’ve altered the true price.

# Adjusted Closing Price and Stock Splits:

A stock split is a corporate action that takes place when a company increases the number of shares in their company. It results in a decrease in the market price of the individual shares and an increase in the number of shares. The stock split can be done in an attempt to lower the price of individual shares for investors. In such a case, the number of shares will increase, and the value of each individual will, in turn, decrease because they will represent a smaller percentage of shares.


# Understanding the Effect of Dividends

Dividends are payouts that a company can distribute to shareholders when stocks and profits are appreciating. A company might pay out a dividend as an award of additional shares to a stockholder or as a cash return. While dividends are good for stockholders, they actually decrease the value of each company stock.

The decrease is caused by the fact that paying out dividends reduces the value of the company because they are transferring money or stocks into the hands of shareholders instead of investing it back into the company. Unlike closing price, adjusted closing price reflects devaluation caused by dividend disbursement.

Overall, the adjusted closing price will give you a better idea of the overall value of the stock and help you make informed decisions about buying and selling, while the closing stock price will tell you the exact cash value of a share of stock at the end of the trading day.

source: https://finance.zacks.com/impact-stock-splits-total-stockholders-equity-3325.html  


# Benefits of the Adjusted Closing Price
The main advantage of adjusted closing prices is that they make it easier to evaluate stock performance. Firstly, the adjusted closing price helps investors understand how much they would have made by investing in a given asset. Most obviously, a 2-for-1 stock split does not cause investors to lose half their money. Since successful stocks often split repeatedly, graphs of their performance would be hard to interpret without adjusted closing prices.

Secondly, the adjusted closing price allows investors to compare the performance of two or more assets. Aside from the clear issues with stock splits, failing to account for dividends tends to understate the profitability of value stocks and dividend growth stocks. Using the adjusted closing price is also essential when comparing the returns of different asset classes over the long term. For example, the prices of high-yield bonds tend to fall in the long run. That does not mean these bonds are necessarily poor investments. Their high yields offset the losses and more, which can be seen by looking at the adjusted closing prices of high-yield bond funds.

*The adjusted closing price provides the most accurate record of returns for long-term investors looking to design asset allocations.*

Source:https://www.investopedia.com/terms/a/adjusted_closing_price.asp 

# Analyzing the highest / lowest percentage of decreasing in closing price of Apple Inc shares / stocks

Let's find the highest percent decrement of closing price from the year 1980 till 2021.

In [21]:
%%sql
/*Highest percent decrement of closing price to adjusted closing price within respective year*/
SELECT Year, MAX(Percent_CP_Decrement)
FROM (SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade,
       ROUND(SUM (Close),2) AS Close_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst,
       ROUND(((SUM(Close)-SUM([Adj Close])) / SUM(Close))*100,2) || ' ' || '%' AS Percent_CP_Decrement
FROM AppleStock
GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('2015', '8.67 %')]

Query above shows 2015 set the highest in percentage of decreasing to adjusted closing price.

Now, the lowest percent decrement of closing price from the year 1980 till 2021.

In [22]:
%%sql
/*Lowest percent decrement of closing price to adjusted closing price within respective year*/
SELECT Year, MIN(Percent_CP_Decrement)
FROM (SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade,
       ROUND(SUM (Close),2) AS Close_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst,
       ROUND(((SUM(Close)-SUM([Adj Close])) / SUM(Close))*100,2) || ' ' || '%' AS Percent_CP_Decrement
FROM AppleStock
GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('2021', '0.29 %')]

Thus, the lowest percentage decreasing of closing price held by the year 2021.

Now, let us compare the Opening Price and Closing Price of Apple Inc shares for each year.

In [14]:
%%sql
/*Use SUM function to compare opening price and adjusted closing price of shares on respective year*/
SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade, ROUND(SUM (Open),2) AS Opening_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst 
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 13, 1.77, 1.38),
 ('1981', 253, 27.55, 21.49),
 ('1982', 253, 21.62, 16.92),
 ('1983', 253, 42.32, 33.12),
 ('1984', 253, 30.35, 23.69),
 ('1985', 252, 22.95, 17.8),
 ('1986', 253, 36.58, 28.69),
 ('1987', 253, 88.0, 68.99),
 ('1988', 253, 93.87, 74.1),
 ('1989', 252, 93.63, 74.73),
 ('1990', 253, 84.71, 68.38),
 ('1991', 253, 118.48, 96.6),
 ('1992', 254, 124.29, 102.13),
 ('1993', 253, 92.76, 76.85),
 ('1994', 252, 76.62, 64.58),
 ('1995', 252, 91.4, 77.74),
 ('1996', 254, 56.8, 48.45),
 ('1997', 253, 40.73, 34.8),
 ('1998', 252, 68.65, 58.96),
 ('1999', 252, 129.73, 111.44),
 ('2000', 252, 206.06, 176.36),
 ('2001', 248, 89.3, 76.77),
 ('2002', 252, 86.08, 73.84),
 ('2003', 252, 83.35, 71.55),
 ('2004', 252, 159.4, 137.07),
 ('2005', 252, 419.67, 360.17),
 ('2006', 251, 636.35, 544.23),
 ('2007', 251, 1150.92, 985.87),
 ('2008', 253, 1285.91, 1099.9),
 ('2009', 252, 1319.57, 1132.86),
 ('2010', 252, 2339.62, 2005.02),
 ('2011', 252, 3276.55, 2808.76),
 ('2012', 250, 5148.68, 4420.39),
 ('2013', 252, 4258.15, 3729.39),
 ('2014', 252, 5809.85, 5219.4),
 ('2015', 252, 7571.07, 6906.77),
 ('2016', 252, 6583.99, 6143.41),
 ('2017', 251, 9440.42, 8972.91),
 ('2018', 251, 11866.74, 11439.36),
 ('2019', 252, 13095.75, 12851.15),
 ('2020', 253, 24102.72, 23898.75),
 ('2021', 236, 32683.51, 32618.07)]

Now, change the shares prices decrement in percentage format.

In [15]:
%%sql
/*Decrement percentage from opening price to adjusted closing price within respective year*/
SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade,
       ROUND(SUM (Open),2) AS Opening_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst,
       ROUND(((SUM(Open)-SUM([Adj Close])) / SUM(Open))*100,2) || ' ' || '%' AS Percent_SharePrice_Decrement
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 13, 1.77, 1.38, '21.83 %'),
 ('1981', 253, 27.55, 21.49, '21.97 %'),
 ('1982', 253, 21.62, 16.92, '21.72 %'),
 ('1983', 253, 42.32, 33.12, '21.73 %'),
 ('1984', 253, 30.35, 23.69, '21.93 %'),
 ('1985', 252, 22.95, 17.8, '22.44 %'),
 ('1986', 253, 36.58, 28.69, '21.56 %'),
 ('1987', 253, 88.0, 68.99, '21.6 %'),
 ('1988', 253, 93.87, 74.1, '21.06 %'),
 ('1989', 252, 93.63, 74.73, '20.19 %'),
 ('1990', 253, 84.71, 68.38, '19.28 %'),
 ('1991', 253, 118.48, 96.6, '18.47 %'),
 ('1992', 254, 124.29, 102.13, '17.83 %'),
 ('1993', 253, 92.76, 76.85, '17.15 %'),
 ('1994', 252, 76.62, 64.58, '15.71 %'),
 ('1995', 252, 91.4, 77.74, '14.95 %'),
 ('1996', 254, 56.8, 48.45, '14.7 %'),
 ('1997', 253, 40.73, 34.8, '14.58 %'),
 ('1998', 252, 68.65, 58.96, '14.12 %'),
 ('1999', 252, 129.73, 111.44, '14.1 %'),
 ('2000', 252, 206.06, 176.36, '14.41 %'),
 ('2001', 248, 89.3, 76.77, '14.03 %'),
 ('2002', 252, 86.08, 73.84, '14.21 %'),
 ('2003', 252, 83.35, 71.55, '14.16 %'),
 ('2004', 252, 159.4, 137.07, '14.01 %'),
 ('2005', 252, 419.67, 360.17, '14.18 %'),
 ('2006', 251, 636.35, 544.23, '14.48 %'),
 ('2007', 251, 1150.92, 985.87, '14.34 %'),
 ('2008', 253, 1285.91, 1099.9, '14.47 %'),
 ('2009', 252, 1319.57, 1132.86, '14.15 %'),
 ('2010', 252, 2339.62, 2005.02, '14.3 %'),
 ('2011', 252, 3276.55, 2808.76, '14.28 %'),
 ('2012', 250, 5148.68, 4420.39, '14.15 %'),
 ('2013', 252, 4258.15, 3729.39, '12.42 %'),
 ('2014', 252, 5809.85, 5219.4, '10.16 %'),
 ('2015', 252, 7571.07, 6906.77, '8.77 %'),
 ('2016', 252, 6583.99, 6143.41, '6.69 %'),
 ('2017', 251, 9440.42, 8972.91, '4.95 %'),
 ('2018', 251, 11866.74, 11439.36, '3.6 %'),
 ('2019', 252, 13095.75, 12851.15, '1.87 %'),
 ('2020', 253, 24102.72, 23898.75, '0.85 %'),
 ('2021', 236, 32683.51, 32618.07, '0.2 %')]

Maximum decrement of share price in which year?

In [24]:
%%sql
/*Maximum decrement percentage from opening price to adjusted closing price within respective year*/
SELECT Year, MAX(Percent_SharePrice_Decrement)
FROM (SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade,
       ROUND(SUM (Open),2) AS Opening_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst,
       ROUND(((SUM(Open)-SUM([Adj Close])) / SUM(Open))*100,2) || ' ' || '%' AS Percent_SharePrice_Decrement
       FROM AppleStock
       GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('2015', '8.77 %')]

Minimum decrement of share price in which year?

In [25]:
%%sql
/*Minimum decrement percentage from opening price to adjusted closing price within respective year*/
SELECT Year, MIN(Percent_SharePrice_Decrement)
FROM (SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade,
       ROUND(SUM (Open),2) AS Opening_Price, ROUND(SUM ([Adj Close]),2) AS Close_Price_Adjst,
       ROUND(((SUM(Open)-SUM([Adj Close])) / SUM(Open))*100,2) || ' ' || '%' AS Percent_SharePrice_Decrement
       FROM AppleStock
       GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('2021', '0.2 %')]

We can see the pattern:

* 2015 is the year which the percent of opening price of Apple Inc shares decreases the most to adjusted closing price, with 8.77%. This year also shows that Apple Inc have increases its stocks splits that cause its decreasing percentage of adjusted closing price to be the dominant among other years.

* However, we can see Apple Inc improved years after 2015, in which 2021 sets the record of having the lowest percent decrement to adjusted closing price as well as lowest percent of opening price to closing price.


So what does actually happened at Apple Inc on the year 2015,which then serves as improvements inspirations for Apple Inc to be the most successful company in the world?

According to the source: https://igotoffer.com/apple/history-apple-2015 

* The are six product lines / servises of Apples that brings impacts to the world as well as progressive technology. Among those are:

     a) 12-inch MacBook
     
     b) 12.9-inch iPad Pro
     
     c) Apple Watch
     
     d) Apple Pay
     
     e) Apple Music
     
     f) Beats 1

* There are some occurences that breaks a few records in Sales history of Apple Inc within the year 2015.


However, there are some errors in 2015 that caused Apple Inc unable to maintain their great quality of product performance. Those events are:

* April 27, 2015: Apple announces that its Board of Directors has authorized an increase of more than 50 percent to the Company’s program to return capital to shareholders. Under the expanded program, Apple plans to utilize a cumulative total of 200 billion dollar of cash by the end of March 2017. As part of the revised program, the Board has increased its share repurchase authorization to 140 billion dollar from the 90 billion dollar level announced last year. In addition, the Company expects to continue to net-share-settle vesting restricted stock units. The Board has also approved an increase of 11 percent to the Company’s quarterly dividend, and has declared a dividend of 0.52 dollar per share, payable on May 14, 2015 to shareholders of record as of the close of business on May 11, 2015. From the inception of its capital return program in August 2012 through March 2015, Apple has returned over 112 billion dollar to shareholders, including 80 billion dollar in share repurchases.

* June 3, 2015: Apple announces a voluntary recall of Beats Pill XL speakers, including a refund for customers. Apple has determined that, in rare cases, the battery in the Beats Pill XL may overheat and pose a fire safety risk. The recall does not affect any other Beats or Apple products. Because customer safety is the company’s top priority, Apple is asking customers to stop using their Beats Pill XL speakers. Customers who purchased a Beats Pill XL speaker have been invited to visit www.apple.com/support/beats-pillxl-recall for details about how to return their product to Apple, and how to receive an Apple Store® credit or electronic payment of 325 dollars. The Beats Pill XL can be identified by the Beats “b” logo on the speaker grille and the words “beats pill XL” on the handle. The product is sold in five colors: black, metallic sky, pink, titanium and white. The Beats Pill XL was introduced by Beats by Dre in November 2013. Apple acquired Beats in 2014.

* July 2015: An Irish firm Probendi sues Apple Inc. for trademark infringement, arguing that through keyword advertising on the Google search engine Apple Watch appears on search results pages when users search for the trademarked term iWatch.

* September 9, 2015: Apple classifies the first generation Apple TV as being obsolete/vintage. This announcement means that owners of first generation Apple TV will no longer be able to get service for their old devices after this date.

* October 1, 2015: Apple starts shipping of Apple TV 4th generation, which uses a new operating system tvOS with an associated app store. The new tvOS allows to download third-party apps for video, audio, games, and other content. These apps and games must must include interfacing with the new touchpad-enabled Siri remote to be used. tvOS is 95% based on iOS 9 with the interface and APIs being tweaked to support the big screen. However, upon launch, there were several issues which included incompatibility with Apple’s Remote app for iOS and WatchOS (see also December 2015).

* December 17, 2015: Apple’s One to One service ends. The service was gradually retired during the year with the final renewals ending on December 17. One to One Membership was a scheme Apple offered that enabled users to get personal training with a new iDevice. For a fee, training typically took place in Apple Retail Stores, in the form of an individual metting with an Apple Genius.

* December 2015: Apple fixes unexpected serious issues with incompatibility of the Apple TV 4th generation with Apple’s Remote app for iOS and WatchOS.

* December 2015: Amazon.com declines to develop an application for the Apple TV as Amazon is in direct competition with Apple with Amazon’s Fire TV. Amazon thus declines to develop an application for the Apple TV and begins to stop selling Apple TVs in their Apple products section of their online store. Amazon chooses to remove all Apple TV SKUs.
     
Additional source to read: https://www.ijser.org/researchpaper/Apple-Inc-Industry-Analysis-Business-Policy-and-Strategy.pdf

Credit source for reading: https://www.investopedia.com/articles/stocks/12/history-apple-stock-increases.asp#toc-the-heavy-hitters-apple-products-that-made-a-difference

# Analyzing Highest / Lowest value of Apple Stock Price

In [16]:
%%sql
/*Use SUM function to find the highest stock price and lowest stock price respective year*/
SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade, ROUND(SUM (Low),2) AS Lowest_Price, ROUND(SUM (High),2) AS Highest_Price 
FROM AppleStock
GROUP BY Year;

 * sqlite:///applestock.db
Done.


[('1980', 13, 1.77, 1.77),
 ('1981', 253, 27.46, 27.64),
 ('1982', 253, 21.41, 21.93),
 ('1983', 253, 41.35, 43.28),
 ('1984', 253, 29.76, 30.94),
 ('1985', 252, 22.66, 23.19),
 ('1986', 253, 35.98, 37.2),
 ('1987', 253, 85.97, 90.02),
 ('1988', 253, 92.37, 95.3),
 ('1989', 252, 92.2, 95.18),
 ('1990', 253, 83.17, 86.33),
 ('1991', 253, 116.35, 120.68),
 ('1992', 254, 122.39, 126.1),
 ('1993', 253, 91.0, 94.37),
 ('1994', 252, 75.18, 78.1),
 ('1995', 252, 89.79, 92.85),
 ('1996', 254, 55.57, 57.65),
 ('1997', 253, 39.82, 41.47),
 ('1998', 252, 67.0, 70.35),
 ('1999', 252, 126.68, 132.97),
 ('2000', 252, 199.69, 212.68),
 ('2001', 248, 86.9, 91.97),
 ('2002', 252, 84.22, 87.85),
 ('2003', 252, 81.93, 84.8),
 ('2004', 252, 157.16, 162.13),
 ('2005', 252, 413.67, 425.71),
 ('2006', 251, 625.8, 644.88),
 ('2007', 251, 1131.15, 1165.99),
 ('2008', 253, 1254.68, 1311.18),
 ('2009', 252, 1304.68, 1336.46),
 ('2010', 252, 2311.63, 2361.32),
 ('2011', 252, 3242.68, 3306.81),
 ('2012', 250, 5088.58, 5194.87),
 ('2013', 252, 4214.22, 4298.75),
 ('2014', 252, 5762.96, 5859.8),
 ('2015', 252, 7488.38, 7638.4),
 ('2016', 252, 6532.49, 6641.91),
 ('2017', 251, 9380.35, 9500.73),
 ('2018', 251, 11745.77, 11984.88),
 ('2019', 252, 12995.11, 13219.4),
 ('2020', 253, 23759.94, 24455.43),
 ('2021', 236, 32365.28, 33022.62)]

From the result query above, in what year marked the ultimate highest price of stocks of Apple Inc?

In [17]:
%%sql
/*Use MAX aggregate function to find highest stock price of Apple history*/
SELECT Year, MAX(Highest_Price)
FROM (SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade, ROUND(SUM (Low),2) AS Lowest_Price, ROUND(SUM (High),2) AS Highest_Price 
FROM AppleStock
GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('2021', 33022.62)]

This shows that Apple Inc recorded to have the highest stock price within the year 2021.

Next, what year marked the ultimate lowest price of stocks of Apple Inc?

In [18]:
%%sql
/*Use MIN aggregate function to find highest stock price of Apple history*/
SELECT Year, MIN(Lowest_Price)
FROM (SELECT strftime('%Y', Date) AS Year, COUNT(*) AS Occur_Shares_Trade, ROUND(SUM (Low),2) AS Lowest_Price, ROUND(SUM (High),2) AS Highest_Price 
FROM AppleStock
GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('1980', 1.77)]

It is logical that Apple Inc has the lowest stock price in the year 1980, bacause this is the starting year that the incorporation is known to public.

# Analyzing the highest / lowest market capitalization of Apple Inc's hares trading history

Querying to find the year which Apple Inc had the highest market capitalization.

In [27]:
%%sql
SELECT Year, MAX(Market_Capitalization_AppleInc)
FROM (SELECT strftime('%Y', Date) AS Year, ROUND(SUM (Volume * [Adj Close]),2) AS Market_Capitalization_AppleInc
FROM AppleStock
GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('2020', 3653041355098.71)]

Apple Inc reached highest market capitalization in the year 2020.

Querying to find the year which Apple Inc had the highest market capitalization.

In [28]:
%%sql
SELECT Year, MIN(Market_Capitalization_AppleInc)
FROM (SELECT strftime('%Y', Date) AS Year, ROUND(SUM (Volume * [Adj Close]),2) AS Market_Capitalization_AppleInc
FROM AppleStock
GROUP BY Year);

 * sqlite:///applestock.db
Done.


[('1980', 138349567.55)]

It is logical that Apple Inc had the lowest market capitalization in the year 1980.

# Insights

* It is true that the year 2015 changed the perspective of human population entirely on how technology works. Humans are getting more innovative in technology progression that ease a lot of our daily activities to be done in quick manner, that helped in cost reduction, as well as increasing time-efficient.

* Inspiration and Innovation are great for building a good future. But, this plan must be examined with all aspects of human safety, before its exposure. Any technology advancement requires a cautious attitude, as it is essential for humans well-being in long-term.

*END OF QUERY*